File: language-support/hs/bindings/src/DA/Ledger/Services/CommandSubmissionService.hs
-- Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
-- SPDX-License-Identifier: Apache-2.0

{-# LANGUAGE GADTs #-}

module DA.Ledger.Services.CommandSubmissionService (submit) where

import Data.Functor
import Com.Daml.Ledger.Api.V1.CommandSubmissionService
import DA.Ledger.Convert (lowerCommands)
import DA.Ledger.GrpcWrapUtils
import DA.Ledger.LedgerService
import DA.Ledger.Types (Commands)
import Google.Protobuf.Empty (Empty(..))
import Network.GRPC.HighLevel.Generated

submit :: Commands -> LedgerService (Either String ())
submit commands =
    makeLedgerService $ \timeout config mdm ->
    withGRPCClient config $ \client -> do
        service <- commandSubmissionServiceClient client
        let CommandSubmissionService rpc = service
        let request = SubmitRequest (Just (lowerCommands commands))
        rpc (ClientNormalRequest request timeout mdm)
            >>= unwrapWithCommandSubmissionFailure
            <&> fmap (\Empty{} -> ())

File: sdk/canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/ledger/api/services/CommandSubmissionService.scala
// Copyright (c) 2025 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.canton.ledger.api.services

import com.digitalasset.canton.ledger.api.messages.command.submission.SubmitRequest
import com.digitalasset.canton.lifecycle.FutureUnlessShutdown
import com.digitalasset.canton.logging.LoggingContextWithTrace

trait CommandSubmissionService {
  def submit(
      request: SubmitRequest
  )(implicit
      loggingContext: LoggingContextWithTrace
  ): FutureUnlessShutdown[Unit]
}
